### PR TITLE
Genotype storage review

### PR DESCRIPTION
**Pull request by lchorbadjiev**
## Background

The initial abstraction for genotype storage was too close to genotype storages we implemented in the beginning - filesystem storage and impala storage (schema1).

Recently we started adding new genotype storage - impala schema2 genotype storage. BigQuery genotype storage and a filesystem genotype storage based on schema2 are planned.

Additionally, we introduced the concept of import storage that works with import project infrastructure and implements operations needed for the import of genotype variants data into the corresponding genotype storage.

The problems that are addressed in this PR are:
* reduce the genotype storage abstraction API - remove simple_study_import, is_impala, and is_filesystem methods from the genotype storages hierarchy;
* make genotype and import storage abstraction more easily extensible;
* address the creation of genotype and import storages;
* prepare integration testing infrastructure for easy test parametrization with genotype storage.

## Aim

The major changes introduced in the PR are:
* Extension points for registering genotype storage and import storage factories are added. The introduction of these extension points allows the definition of genotype/import storage in external projects.  
* Configuration validation is now the responsibility of each individual genotype storage. Common genotype storage configuration schema is reduced to require `id` and `storage_type` parameters.
*  Filesystem import storage is created that allows using import project infrastructure to import genotype variants data into filesystem genotype storage.
* The `simple_study_import` method of genotype storage is removed. To ensure this aim we: 
   * changed the tool `simple_study_import.py` to use import project infrastructure.
   * schema1 tests are changed to use import project infrastructure.
* Packages related to the genotype/import storages are reorganized. This reorganization probably should be tuned further while working on BigQuery and filesystem2 storages.
* Some Infrastructure for test parametrization with genotype storage is added. Some integration tests for genotype data variants queries are migrated to this infrastructure.

## Implementation

The definition of genotype storage requires the definition of two classes - one for genotype storage and one for import storage. Some experiments to merge these two abstractions were tried but not successful. 

At the moment there are two extension points: one for registering genotype storage factories and one for registering import storage factories. This is something that feels awkward to me but I decided to address it in a separate PR.

Parametrization of integration tests uses pytest_generate_tests to inject genotype_storage fixture. I am not sure if there is a better way to do that.

### CODE REVIEW

File: dae/dae/configuration/tests/conftest.py [hunk 1]
```diff
@@ -1,7 +1,8 @@
 import os
 import pytest
 
-from dae.configuration.gpf_config_parser import validate_path
+from dae.configuration.utils import \
+    validate_path
```
Comment (svetlin-mladenov) [inline dae/dae/configuration/tests/conftest.py:5]: Why have you split this line? It is shorter than 80.

File: dae/dae/impala_storage/schema1/impala_variants.py [hunk 2]
```diff
@@ -728,49 +729,49 @@ def _check_summary_variants_table(self):
...
-                COUNT(
-                    DISTINCT
-                        bucket_index,
-                        summary_variant_index,
-                        family_variant_index
-                )
-            FROM {self.db}.{self.variants_table}
-            {where_clause}
-        """
+    # def build_count_query(
```
Comment (svetlin-mladenov) [inline dae/dae/impala_storage/schema1/impala_variants.py:732]: If not used remove.

File: dae/dae/impala_storage/schema1/tests/test_impala_storage_config.py [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    config = {
+        "id": "genotype_impala",
+        "storage_type": "impala",
+        "hdfs": {
+            "base_dir": "/tmp/test_data",
+            "host": "localhost",
+            "port": 8020,
+            "replication": 1,
+        },
+        # "impala": {
```
Comment (svetlin-mladenov) [inline dae/dae/impala_storage/schema1/tests/test_impala_storage_config.py:272]: Remove if unused.

File: dae/conftest.py [hunk 4]
```diff
@@ -1,10 +1,44 @@
 # pylint: disable=W0621,C0114,C0116,W0212,W0613
-
 import pytest
 
 pytest_plugins = ["dae_conftests.dae_conftests"]
 
 
+# def pytest_sessionfinish(session, exitstatus):
```
[the remark this answers is not in the document]
Comment (lchorbadjiev) [reply dae/conftest.py:7]: fixed

File: dae/dae/configuration/tests/conftest.py (hunk 1 above)
Comment (lchorbadjiev) [reply dae/dae/configuration/tests/conftest.py:5]: fixed

File: dae/dae/impala_storage/schema1/impala_variants.py (hunk 2 above)
Comment (lchorbadjiev) [reply dae/dae/impala_storage/schema1/impala_variants.py:732]: fixed

File: dae/dae/impala_storage/schema1/tests/test_impala_storage_config.py (hunk 3 above)
Comment (lchorbadjiev) [reply dae/dae/impala_storage/schema1/tests/test_impala_storage_config.py:272]: fixed

File: dae_conftests/dae_conftests/dae_conftests.py [hunk 5]
```diff
@@ -411,35 +414,35 @@ def from_prefix_vcf(prefix):
...
     fullpath = relative_to_this_test_folder("fixtures/dae_denovo/denovo")
     config = from_prefix_denovo(fullpath)
-    return config.denovo
+    return config
 
 
-@pytest.fixture
-def dae_denovo(
-        dae_denovo_config, gpf_instance_2013, annotation_pipeline_internal):
+# @pytest.fixture
```
[the remark this answers is not in the document]
Comment (lchorbadjiev) [reply dae_conftests/dae_conftests/dae_conftests.py:427]: fixed

File: dae/dae/gpf_instance/gpf_instance.py [hunk 6]
```diff
@@ -86,8 +81,12 @@ def __init__(
 
     @property  # type: ignore
     @cached
-    def reference_genome(self) -> ReferenceGenome:
-        """Returns reference genome defined in the GPFInstance config"""
+    def reference_genome(self):
+        """Return reference genome defined in the GPFInstance config."""
+        # pylint: disable=import-outside-toplevel
+        from dae.genomic_resources.reference_genome import \
```
Comment (IvoTod) [inline dae/dae/gpf_instance/gpf_instance.py:87]: I feel like we discussed this but I may have forgotten, why do we use inline imports in these cases?

**Review by IvoTod — APPROVED** [review]

File: dae/dae/docs/genotype_storage.rst [hunk 7]
```diff
@@ -0,0 +1,55 @@
+Genotype Storage
+================
+
+To store and query genomic variants we use a genotype storage.
+
+There are two interfaces that define a genotype storage:
+
+* :class:`dae.genotype_storage.genotype_storage.GenotypeStorage`  that defines
+  how to use the storage for quering genomic variants;
```
[the remark this answers is not in the document]
Comment (lchorbadjiev) [reply dae/dae/docs/genotype_storage.rst:9]: fixed

File: dae/dae/docs/genotype_storage.rst [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+
+When you want to create a new genotype storage you have to create two classes - 
+a genotype storage class that inherits the 
+:class:`dae.genotype_storage.genotype_storage.GenotypeStorage`
+and an import storage class that inherits 
+:class:`dae.import_tools.import_tools.ImportStorage`. These classes should be
+registered in two extenstion points, defined in the `setup.py`:
+
+* for genotype storage class use `[dae.genotype_storage.factories]` extenstion
```
[the remark this answers is not in the document]
Comment (lchorbadjiev) [reply dae/dae/docs/genotype_storage.rst:22]: fixed

File: dae/dae/docs/genotype_storage.rst [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  point;
+
+* for import storage class use `[dae.import_tools.storages]` extenstion point.
```
[the remark this answers is not in the document]
Comment (lchorbadjiev) [reply dae/dae/docs/genotype_storage.rst:25]: fixed

**Pull request merged**